pkg/segment/segment.go

```diff
@@ -0,0 +1,153 @@
+package segment
+
+import (
+	"errors"
+	"fmt"
+	"github.com/openshift/odo/pkg/preference"
+	"github.com/openshift/odo/pkg/version"
+	"github.com/pborman/uuid"
+	"golang.org/x/term"
+	"gopkg.in/segmentio/analytics-go.v3"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+// Writekey will be the API key used to send data to the correct source on Segment
+var WriteKey = "4xGV1HV7K2FtUWaoAozSBD7SNCBCJ65U"
+
+type Client struct {
+	segmentClient     analytics.Client
+	config            *preference.PreferenceInfo
+	telemetryFilePath string
+}
+
+func NewClient(config *preference.PreferenceInfo) (*Client, error) {
+	homeDir, _ := os.UserHomeDir()
+	return newCustomClient(config,
+		filepath.Join(homeDir, ".redhat", "anonymousId"),
+		analytics.DefaultEndpoint,
+	)
+}
+
+func newCustomClient(config *preference.PreferenceInfo, telemetryFilePath string, segmentEndpoint string) (*Client, error) {
+	client, err := analytics.NewWithConfig(WriteKey, analytics.Config{
+		Endpoint: segmentEndpoint,
+		DefaultContext: &analytics.Context{
+			IP: net.IPv4(0, 0, 0, 0),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		segmentClient:     client,
+		config:            config,
+		telemetryFilePath: telemetryFilePath,
+	}, nil
+}
+
+// Close client connection and send the data
+func (c *Client) Close() error {
+	return c.segmentClient.Close()
+}
+
+// Upload prepares the data to be sent to segment and send it once the client connection closes
+func (c *Client) Upload(action string, duration time.Duration, err error) error {
+	// if the user has not consented for telemetry, return
+	if !c.config.GetConsentTelemetry() {
```

Comment:
I would suggest we refactor this consent piece in a way that we dont even initialise the segment client if the consent is not set, and lets do that in the genericclioptions? would save us some initialisation time?

Reply:
Right, that sounds like a good idea. But I think I would like to keep this check here, in case someone explicitly calls the Upload function, I understand the chances of that are slim, but I would still keep it.

Reply:
yeah I dont mind if this code stays here - but we should only init the client if consent is true in the beginning itself